backend/src/lib.rs

```diff
@@ -1,104 +1,186 @@
+    fn prove(
+        &self,
+        pil: &Analyzed<F>,
+        fixed: &[(&str, Vec<F>)],
+        witness: Option<&[(&str, Vec<F>)]>,
+        prev_proof: Option<Proof>,
+        output_dir: &Path,
+    ) -> io::Result<()> {
+        self.0.prove(pil, fixed, witness, prev_proof, output_dir)
     }
-}
 
-#[cfg(feature = "halo2")]
-impl ProverAggregationWithParams for Halo2AggregationBackend {
-    fn prove<T: FieldElement, R1: io::Read, R2: io::Read>(
-        pil: &Analyzed<T>,
-        fixed: Vec<(&str, Vec<T>)>,
-        witness: Vec<(&str, Vec<T>)>,
-        proof: R1,
-        params: R2,
-    ) -> Proof {
-        halo2::prove_aggr_read_proof_params(pil, fixed, witness, proof, params)
+    fn write_setup(&self, output: &mut dyn io::Write) -> Result<(), Error> {
+        Ok(self.0.write_setup(output)?)
     }
 }
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("input/output error")]
+    IO(#[from] std::io::Error),
+    #[error("operation is not applicable to backend type")]
+    NotApplicable,
+}
+
+pub type Proof = Vec<u8>;
+
+/// Dynamic interface for a backend.
+pub trait Backend<F: FieldElement> {
+    /// Perform the backend proving. If prev_proof is provided, proof
+    /// aggregation is performed.
+    fn prove(
+        &self,
+        pil: &Analyzed<F>,
+        fixed: &[(&str, Vec<F>)],
+        witness: Option<&[(&str, Vec<F>)]>,
+        prev_proof: Option<Proof>,
+        output_dir: &Path,
+    ) -> io::Result<()>;
+
+    /// Write the prover setup to a file, so that it can be loaded later.
+    fn write_setup(&self, output: &mut dyn io::Write) -> Result<(), Error>;
+}
+
+/// Dynamic interface for a backend factory.
+pub trait BackendFactory<F: FieldElement> {
+    /// Maybe perform the setup, and create a new backend object.
+    fn new_from_params(&self, degree: DegreeType) -> Box<dyn Backend<F>>;
+
+    /// Create a backend object from a prover setup loaded from a file.
+    fn load_setup(&self, input: &mut dyn io::Read) -> Result<Box<dyn Backend<F>>, Error>;
+}
+
+/// Trait implemented by all backends.
+trait BackendImpl<F: FieldElement> {
+    fn prove(
+        &self,
+        pil: &Analyzed<F>,
+        fixed: &[(&str, Vec<F>)],
+        witness: Option<&[(&str, Vec<F>)]>,
+        prev_proof: Option<Proof>,
+        output_dir: &Path,
+    ) -> io::Result<()>;
+}
+
+/// Trait implemented by backends that have a setup phase that must be saved to
+/// a file.
+trait BackendWithSetup<F: FieldElement>
+where
+    Self: Sized + BackendImpl<F>,
+{
+    /// Perform the setup and create a new backend object.
+    fn new_setup_from_params(degree: DegreeType) -> Self;
+
+    /// Create a backend object from a setup loaded from a file.
+    fn load_setup(input: &mut dyn io::Read) -> Result<Self, io::Error>;
+
+    /// Write the setup to a file.
+    fn write_setup(&self, output: &mut dyn io::Write) -> Result<(), io::Error>;
```

Comment:
I don't think this should exist. I think the client should consume the struct member and just write to whatever it wants to write it

Reply:
Didn't understand. `write_setup` is how a setup is written to a file, to be loaded later with `load_setup`

Reply:
I don't think it should be "loaded later". If the setup params are created with `new`, the params are already available as a field in the struct. The client can just write it to wherever it wants

Reply:
the client can write wherever it wants, by calling `write_setup()`